File: src/parsers/linePrefix.test.ts
/**
 * Copyright 2025 IBM Corp.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { LinePrefixParser, LinePrefixParserError } from "@/parsers/linePrefix.js";
import { z } from "zod";
import { JSONParserField, ZodParserField } from "@/parsers/field.js";
import { splitString } from "@/internals/helpers/string.js";
import { ValueError } from "@/errors.js";

describe("LinePrefixParser", () => {
  it.each([1, 5, 20, 50, 100, 1000])("Handles arbitrary chunk size", async (chunkSize) => {
    const parser = new LinePrefixParser({
      thought: {
        prefix: "Thought:",
        next: ["tool_name", "final_answer"],
        isStart: true,
        field: new ZodParserField(z.string().min(1)),
      },
      tool_name: {
        prefix: "Tool Name:",
        next: ["tool_caption"],
        field: new ZodParserField(z.enum(["Google"])),
      },
      tool_caption: {
        prefix: "Tool Caption:",
        next: ["tool_input"],
        field: new ZodParserField(z.string().min(1)),
      },
      tool_input: {
        prefix: "Tool Input:",
        next: ["tool_output"],
        field: new JSONParserField({
          schema: z.record(z.string()),
          base: {},
        }),
      },
      tool_output: {
        prefix: "Tool Output:",
        next: ["final_answer"],
        isEnd: true,
        field: new JSONParserField({
          schema: z.record(z.string()),
          base: {},
        }),
      },
      final_answer: {
        prefix: "Final Answer:",
        next: [],
        isStart: true,
        isEnd: true,
        field: new ZodParserField(z.string().min(1)),
      },
    } as const);

    // Gather all update events and then compare if they equal the same
    const updates = {
      partial: {},
      final: {},
    };
    parser.emitter.on("update", async ({ key, field }) => {
      Object.assign(updates.final, { [key]: field.raw });
    });
    parser.emitter.on("partialUpdate", async ({ key, field }) => {
      Object.assign(updates.partial, { [key]: field.raw });
    });

    const chunks = splitString(
      `Thought: I don't know who is president of USA.
But I can use GoogleSearch to find out.
Tool Name: Google
Tool Caption: Searching for USA president
Tool Input: {"query":"USA president"}
Tool Output: {"answer":"Joe Biden"}
Final Answer: Joe Biden is the current president of USA.`,
      { size: chunkSize, overlap: 0 },
    );
    for (const chunk of chunks) {
      await parser.add(chunk);
    }
    const state = await parser.end();

    expect(state).toStrictEqual({
      thought: `I don't know who is president of USA.
But I can use GoogleSearch to find out.`,
      tool_name: "Google",
      tool_caption: "Searching for USA president",
      tool_input: { query: "USA president" },
      tool_output: { answer: "Joe Biden" },
      final_answer: "Joe Biden is the current president of USA.",
    });
    expect(updates.partial).toStrictEqual(updates.final);
  });

  describe("Streaming", () => {
    it("JSON", async () => {
      const parser = new LinePrefixParser({
        start: {
          prefix: "Start:",
          field: new JSONParserField({
            schema: z.object({
              a: z.number().int(),
              b: z.number().int(),
            }),
            base: {},
          }),
          next: [],
          isStart: true,
          isEnd: true,
        },
      } as const);
      await parser.add("Start:");

      for (const field of [
        {
          value: `{"a`,
          expected: {},
        },
        {
          value: `":1,`,
          expected: { a: 1 },
        },
        {
          value: `"b`,
          expected: { a: 1 },
        },
        {
          value: `":`,
          expected: { a: 1, b: undefined },
        },
        {
          value: `2,`,
          expected: { a: 1, b: 2 },
        },
      ]) {
        await Promise.all([
          new Promise<void>((resolve) => {
            parser.emitter.on(
              "partialUpdate",
              (update) => {
                expect(update.delta).toStrictEqual(field.value);
                expect(update.field.getPartial()).toStrictEqual(field.expected);
                resolve();
              },
              {
                once: true,
                isBlocking: true,
              },
            );
          }),
          parser.add(field.value),
        ]);
      }
      await parser.end();
    });
  });

  describe("Fallback", () => {
    const getParser = () =>
      new LinePrefixParser(
        {
          thought: {
            prefix: "Thought:",
            field: new ZodParserField(z.string()),
            isStart: true,
            next: ["final_answer"],
          },
          final_answer: {
            prefix: "Final Answer:",
            field: new ZodParserField(z.string()),
            isEnd: true,
            next: [],
          },
        },
        {
          fallback: (value) =>
            [
              { key: "thought", value: "I now know the final answer." },
              {
                key: "final_answer",
                value,
              },
            ] as const,
        },
      );

    it("Process", async () => {
      const parser = getParser();
      await parser.add("2+2=4");
      await expect(parser.end()).resolves.toMatchInlineSnapshot(`
        {
          "final_answer": "2+2=4",
          "thought": "I now know the final answer.",
        }
      `);
    });

    it("Process #2", async () => {
      const parser = getParser();
      await parser.add("A\nB\nC");
      await expect(parser.end()).resolves.toMatchInlineSnapshot(`
        {
          "final_answer": "A
        B
        C",
          "thought": "I now know the final answer.",
        }
      `);
    });

    it("Process #3", async () => {
      const parser = getParser();
      await parser.add("Thought");
      await expect(parser.end()).resolves.toMatchInlineSnapshot(`
        {
          "final_answer": "Thought",
          "thought": "I now know the final answer.",
        }
      `);
    });
  });

  describe("Edge cases", () => {
    it("Prevents processing when line starts with a potential prefix", async () => {
      const parser = new LinePrefixParser({
        final_answer: {
          prefix: "Final Answer:",
          field: new ZodParserField(z.string()),
          isStart: true,
          isEnd: true,
          next: ["final_detail"],
        },
        final_detail: {
          prefix: "Final Detail:",
          field: new ZodParserField(z.string()),
          isStart: true,
          isEnd: true,
          next: ["final_answer"],
        },
      });

      const deltas: string[] = [];
      parser.emitter.on("partialUpdate", ({ key, delta }) => {
        expect(key).toStrictEqual("final_answer");
        deltas.push(delta);
      });
      await parser.add("Final Answer: 4\n");
      await parser.add("\n"); // we can stream for multiple lines
      expect(deltas.join("")).toStrictEqual("4\n");
      expect(deltas.length).toBe(2);
      await parser.add("Final ");
      expect(deltas.length).toBe(2);
      await parser.add("Detail");
      expect(deltas.length).toBe(2);
      await parser.end();
      expect(deltas.join("")).toStrictEqual("4\n\nFinal Detail");
      expect(deltas.length).toBe(3);
    });

    it.each([true, false])("Correctly interprets new lines", async (shouldThrow) => {
      const input = `Thought: Summarize the discussion. Final Answer: The discussion thread is about ...`;
      const parser = new LinePrefixParser({
        thought: {
          prefix: "Thought:",
          next: ["final_answer"],
          isEnd: !shouldThrow,
          isStart: true,
          field: new ZodParserField(z.string().min(1)),
        },
        final_answer: {
          prefix: "Final Answer:",
          next: [],
          isStart: true,
          isEnd: true,
          field: new ZodParserField(z.string().min(1)),
        },
      } as const);

      const chunks = splitString(input, { size: 25, overlap: 0 });
      for (const chunk of chunks) {
        await parser.add(chunk);
      }

      if (shouldThrow) {
        await expect(parser.end()).rejects.toThrowError(LinePrefixParserError);
      } else {
        await parser.end();
        expect(parser.finalState).toMatchObject({
          thought: `Summarize the discussion. Final Answer: The discussion thread is about ...`,
        });
      }
    });

    it("Ignores unrelated text and non-starting nodes", async () => {
      const parser = new LinePrefixParser(
        {
          first: {
            prefix: "First:",
            field: new ZodParserField(z.string()),
            isStart: true,
            isEnd: false,
            next: ["second"],
          },
          second: {
            prefix: "Second:",
            field: new ZodParserField(z.string()),
            isStart: false,
            isEnd: true,
            next: [],
          },
        },
        {
          waitForStartNode: true,
        },
      );

      await parser.add("      Random text\nthat\nshould be ignored.\n");
      await parser.add("  Second: This should be ignored too.\n");
      await parser.add("First: first\n");
      await parser.add("Second: second\n");
      await parser.end();
      expect(parser.finalState).toMatchInlineSnapshot(`
        {
          "first": "first",
          "second": "second
        ",
        }
      `);
    });

    it.each([true, false])(
      "Ignores other prefixes when the current node is a termination node",
      async (endOnRepeat) => {
        const parser = new LinePrefixParser(
          {
            thought: {
              prefix: "Thought:",
              field: new ZodParserField(z.string()),
              isStart: true,
              isEnd: true,
              next: ["answer"],
            },
            answer: {
              prefix: "Answer:",
              field: new ZodParserField(z.string()),
              isStart: true,
              isEnd: true,
              next: [],
            },
          },
          {
            endOnRepeat,
          },
        );

        await parser.add(`Answer: The answer is 39, see the intermediate steps that I did:\n`);
        await parser.add(`Thought: 3*(2+5+6)=? can be rewritten as 3*13\n`);
        await parser.add(`Analyze: 3*13=39\n`);
        await parser.add(`Outcome: 39`);
        await parser.end();
        expect(parser.finalState).toStrictEqual({
          answer: `The answer is 39, see the intermediate steps that I did:
Thought: 3*(2+5+6)=? can be rewritten as 3*13
Analyze: 3*13=39
Outcome: 39`,
        });
      },
    );

    it("Premature ends when existing node visited", async () => {
      const parser = new LinePrefixParser(
        {
          thought: {
            prefix: "Thought:",
            field: new ZodParserField(z.string()),
            isStart: true,
            isEnd: false,
            next: ["tool_input", "final_answer"],
          },
          tool_input: {
            prefix: "Tool Input:",
            field: new ZodParserField(z.string()),
            isStart: false,
            isEnd: true,
            next: ["tool_output"],
          },
          tool_output: {
            prefix: "Tool Output:",
            field: new ZodParserField(z.string()),
            isStart: false,
            isEnd: true,
            next: [],
          },
          final_answer: {
            prefix: "Final Answer:",
            field: new ZodParserField(z.string()),
            isStart: false,
            isEnd: true,
            next: [],
          },
        },
        {
          endOnRepeat: true,
        },
      );

      await parser.add(
        [`Thought: Hello!`, `Tool Input: {}`, `Thought: This is ignored.`].join("\n"),
      );
      expect(parser.isDone).toBe(true);
      expect(parser.finalState).toMatchInlineSnapshot(`
        {
          "thought": "Hello!",
          "tool_input": "{}",
        }
      `);
    });
  });

  describe("Error handling", () => {
    it("Throws when no node is provided", () => {
      expect(() => new LinePrefixParser({})).toThrowError(ValueError);
    });

    it("Throws when no start node is provided", () => {
      expect(
        () =>
          new LinePrefixParser({
            a: {
              prefix: "A:",
              isStart: false,
              isEnd: true,
              next: [],
              field: new ZodParserField(z.string()),
            },
          }),
      ).toThrowErrorMatchingInlineSnapshot(`ValueError: At least one start node must be provided!`);
    });

    it("Throws when no start node is provided", () => {
      expect(
        () =>
          new LinePrefixParser({
            a: {
              prefix: "A:",
              isStart: true,
              isEnd: false,
              next: [],
              field: new ZodParserField(z.string()),
            },
          }),
      ).toThrowErrorMatchingInlineSnapshot(`ValueError: At least one end node must be provided!`);
    });

    it("Throws for non-existing transition", () => {
      expect(
        () =>
          new LinePrefixParser({
            a: {
              prefix: "A:",
              isStart: true,
              isEnd: true,
              next: ["b"] as any[],
              field: new ZodParserField(z.string()),
            },
          }),
      ).toThrowErrorMatchingInlineSnapshot(
        `ValueError: Node 'a' contains a transition to non-existing node 'b'.`,
      );
    });

    it("Throws when self-reference", () => {
      expect(
        () =>
          new LinePrefixParser({
            a: {
              prefix: "A:",
              isStart: true,
              isEnd: true,
              next: ["a"],
              field: new ZodParserField(z.string()),
            },
          }),
      ).toThrowErrorMatchingInlineSnapshot(`ValueError: Node 'a' cannot point to itself.`);
    });
  });
});

File: python/tests/parsers/line_prefix_test.py
# Copyright 2025 IBM Corp.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from enum import StrEnum

import pytest

from beeai_framework.emitter import EventMeta
from beeai_framework.parsers.field import ParserField
from beeai_framework.parsers.line_prefix import (
    LinePrefixParser,
    LinePrefixParserError,
    LinePrefixParserNode,
    LinePrefixParserOptions,
    LinePrefixParserUpdate,
)
from beeai_framework.utils.strings import split_string


@pytest.mark.asyncio
@pytest.mark.parametrize("chunk_size", [1, 5, 20, 50, 100, 1000])
async def test_handles_arbitrary_chunk_size(chunk_size: int) -> None:
    class ToolNames(StrEnum):
        Google = "Google"

    config = {
        "thought": LinePrefixParserNode(
            prefix="Thought:", next=["tool_name", "final_answer"], is_start=True, field=ParserField.from_type(str)
        ),
        "tool_name": LinePrefixParserNode(
            prefix="Tool Name:",
            next=["tool_caption"],
            field=ParserField.from_type(ToolNames),
        ),
        "tool_caption": LinePrefixParserNode(
            prefix="Tool Caption:", next=["tool_input"], field=ParserField.from_type(str)
        ),
        "tool_input": LinePrefixParserNode(
            prefix="Tool Input:", next=["tool_output"], field=ParserField.from_type(dict, default={})
        ),
        "tool_output": LinePrefixParserNode(
            prefix="Tool Output:",
            next=["final_answer"],
            is_end=True,
            field=ParserField.from_type(dict, default={}),
        ),
        "final_answer": LinePrefixParserNode(
            prefix="Final Answer:", next=[], is_start=True, is_end=True, field=ParserField.from_type(str)
        ),
    }
    parser = LinePrefixParser(config)

    # Capture update events.
    updates: dict = {"partial": {}, "final": {}}

    def on_update(data: LinePrefixParserUpdate, event: EventMeta) -> None:
        updates["final"][data.key] = data.field.raw

    def on_partial_update(data: LinePrefixParserUpdate, event: EventMeta) -> None:
        updates["partial"][data.key] = data.field.raw

    parser.emitter.on("update", on_update)
    parser.emitter.on("partial_update", on_partial_update)

    text = (
        "Thought: I don't know who is president of USA.\n"
        "But I can use GoogleSearch to find out.\n"
        "Tool Name: Google\n"
        "Tool Caption: Searching for USA president\n"
        'Tool Input: {"query":"USA president"}\n'
        'Tool Output: {"answer":"Joe Biden"}\n'
        "Final Answer: Joe Biden is the current president of USA."
    )
    chunks = split_string(text, size=chunk_size, overlap=0)
    for chunk in chunks:
        await parser.add(chunk)
    state = await parser.end()

    expected_state = {
        "thought": "I don't know who is president of USA.\nBut I can use GoogleSearch to find out.",
        "tool_name": "Google",
        "tool_caption": "Searching for USA president",
        "tool_input": {"query": "USA president"},
        "tool_output": {"answer": "Joe Biden"},
        "final_answer": "Joe Biden is the current president of USA.",
    }
    assert state == expected_state
    assert updates["partial"] == updates["final"]


def get_fallback_parser() -> LinePrefixParser:
    config = {
        "thought": LinePrefixParserNode(
            prefix="Thought:",
            field=ParserField.from_type(str),
            is_start=True,
            next=["final_answer"],
        ),
        "final_answer": LinePrefixParserNode(
            prefix="Final Answer:",
            field=ParserField.from_type(str),
            is_end=True,
            next=[],
        ),
    }

    def fallback(value: str) -> list[dict]:
        return [
            {"key": "thought", "value": "I now know the final answer."},
            {"key": "final_answer", "value": value},
        ]

    return LinePrefixParser(config, options=LinePrefixParserOptions(fallback=fallback))


@pytest.mark.asyncio
async def test_fallback_process() -> None:
    parser = get_fallback_parser()
    await parser.add("2+2=4")
    state = await parser.end()
    expected = {
        "thought": "I now know the final answer.",
        "final_answer": "2+2=4",
    }
    assert state == expected


@pytest.mark.asyncio
async def test_fallback_process_2() -> None:
    parser = get_fallback_parser()
    await parser.add("A\nB\nC")
    state = await parser.end()
    expected = {
        "thought": "I now know the final answer.",
        "final_answer": "A\nB\nC",
    }
    assert state == expected


@pytest.mark.asyncio
async def test_fallback_process_3() -> None:
    parser = get_fallback_parser()
    await parser.add("Thought")
    state = await parser.end()
    expected = {
        "thought": "I now know the final answer.",
        "final_answer": "Thought",
    }
    assert state == expected


@pytest.mark.asyncio
async def test_prevents_processing_potential_prefix() -> None:
    config = {
        "final_answer": LinePrefixParserNode(
            prefix="Final Answer:",
            field=ParserField.from_type(str),
            is_start=True,
            is_end=True,
            next=["final_detail"],
        ),
        "final_detail": LinePrefixParserNode(
            prefix="Final Detail:",
            field=ParserField.from_type(str),
            is_start=True,
            is_end=True,
            next=["final_answer"],
        ),
    }
    parser = LinePrefixParser(config)
    deltas = []

    async def on_partial_update(data: LinePrefixParserUpdate, event: EventMeta) -> None:
        # Only updates for the "final_answer" node are expected.
        assert data.key == "final_answer"
        deltas.append(data.delta)

    parser.emitter.on("partial_update", on_partial_update)
    await parser.add("Final Answer: 4\n")
    await parser.add("\n")
    assert "".join(deltas) == "4\n"
    assert len(deltas) == 2

    await parser.add("Final ")
    assert len(deltas) == 2
    await parser.add("Detail")
    assert len(deltas) == 2
    await parser.end()
    assert "".join(deltas) == "4\n\nFinal Detail"
    assert len(deltas) == 3


@pytest.mark.asyncio
@pytest.mark.parametrize("should_throw", [True, False])
async def test_interprets_new_lines(should_throw: bool) -> None:
    input_str = "Thought: Summarize the discussion. Final Answer: The discussion thread is about ..."
    config = {
        "thought": LinePrefixParserNode(
            prefix="Thought:",
            next=["final_answer"],
            is_start=True,
            is_end=not should_throw,  # invert is_end based on the parameter
            field=ParserField.from_type(str),
        ),
        "final_answer": LinePrefixParserNode(
            prefix="Final Answer:",
            next=[],
            is_start=True,
            is_end=True,
            field=ParserField.from_type(str),
        ),
    }
    parser = LinePrefixParser(config)
    chunks = split_string(input_str, size=25, overlap=0)
    for chunk in chunks:
        await parser.add(chunk)

    if should_throw:
        with pytest.raises(LinePrefixParserError):
            await parser.end()
    else:
        await parser.end()
        expected = "Summarize the discussion. Final Answer: The discussion thread is about ..."
        assert parser.final_state["thought"] == expected


@pytest.mark.asyncio
async def test_ignores_unrelated_text_and_non_starting_nodes() -> None:
    config = {
        "first": LinePrefixParserNode(
            prefix="First:",
            field=ParserField.from_type(str),
            is_start=True,
            is_end=False,
            next=["second"],
        ),
        "second": LinePrefixParserNode(
            prefix="Second:",
            field=ParserField.from_type(str),
            is_start=False,
            is_end=True,
            next=[],
        ),
    }
    # Here we pass the config as a dict of node instances.
    parser = LinePrefixParser(config, options=LinePrefixParserOptions(wait_for_start_node=True))
    await parser.add("      Random text\nthat\nshould be ignored.\n")
    await parser.add("  Second: This should be ignored too.\n")
    await parser.add("First: first\n")
    await parser.add("Second: second\n")
    await parser.end()
    expected = {
        "first": "first",
        "second": "second\n",
    }
    assert parser.final_state == expected


@pytest.mark.asyncio
@pytest.mark.parametrize("end_on_repeat", [True, False])
async def test_ignores_other_prefixes(end_on_repeat: bool) -> None:
    config = {
        "thought": LinePrefixParserNode(
            prefix="Thought:",
            field=ParserField.from_type(str),
            is_start=True,
            is_end=True,
            next=["answer"],
        ),
        "answer": LinePrefixParserNode(
            prefix="Answer:",
            field=ParserField.from_type(str),
            is_start=True,
            is_end=True,
            next=[],
        ),
    }
    parser = LinePrefixParser(config, options=LinePrefixParserOptions(end_on_repeat=end_on_repeat))
    await parser.add("Answer: The answer is 39, see the intermediate steps that I did:\n")
    await parser.add("Thought: 3*(2+5+6)=? can be rewritten as 3*13\n")
    await parser.add("Analyze: 3*13=39\n")
    await parser.add("Outcome: 39")
    await parser.end()
    expected = {
        "answer": (
            "The answer is 39, see the intermediate steps that I did:\n"
            "Thought: 3*(2+5+6)=? can be rewritten as 3*13\n"
            "Analyze: 3*13=39\n"
            "Outcome: 39"
        ),
    }
    assert parser.final_state == expected


@pytest.mark.asyncio
async def test_premature_ends_when_existing_node_visited() -> None:
    config = {
        "thought": LinePrefixParserNode(
            prefix="Thought:",
            field=ParserField.from_type(str),
            is_start=True,
            is_end=False,
            next=["tool_input", "final_answer"],
        ),
        "tool_input": LinePrefixParserNode(
            prefix="Tool Input:",
            field=ParserField.from_type(str),
            is_start=False,
            is_end=True,
            next=["tool_output"],
        ),
        "tool_output": LinePrefixParserNode(
            prefix="Tool Output:",
            field=ParserField.from_type(str),
            is_start=False,
            is_end=True,
            next=[],
        ),
        "final_answer": LinePrefixParserNode(
            prefix="Final Answer:",
            field=ParserField.from_type(str),
            is_start=False,
            is_end=True,
            next=[],
        ),
    }
    parser = LinePrefixParser(config, options=LinePrefixParserOptions(end_on_repeat=True))
    text = "\n".join(["Thought: Hello!", "Tool Input: {}", "Thought: This is ignored."])
    await parser.add(text)
    assert parser.done is True
    expected = {
        "thought": "Hello!",
        "tool_input": "{}",
    }
    assert parser.final_state == expected


def test_throws_when_no_node_provided() -> None:
    with pytest.raises(ValueError):
        LinePrefixParser({})


def test_throws_when_no_start_node_provided() -> None:
    with pytest.raises(ValueError) as excinfo:
        LinePrefixParser(
            {
                "a": LinePrefixParserNode(
                    prefix="A:",
                    is_start=False,
                    is_end=True,
                    next=[],
                    field=ParserField.from_type(str),
                )
            }
        )
    assert "At least one start node must be provided!" in str(excinfo.value)


def test_throws_when_no_end_node_provided() -> None:
    with pytest.raises(ValueError) as excinfo:
        LinePrefixParser(
            {
                "a": LinePrefixParserNode(
                    prefix="A:",
                    is_start=True,
                    is_end=False,
                    next=[],
                    field=ParserField.from_type(str),
                )
            }
        )
    assert "At least one end node must be provided!" in str(excinfo.value)


@pytest.mark.unit
def test_throws_for_non_existing_transition() -> None:
    with pytest.raises(ValueError) as excinfo:
        LinePrefixParser(
            {
                "a": LinePrefixParserNode(
                    prefix="A:",
                    is_start=True,
                    is_end=True,
                    next=["b"],
                    field=ParserField.from_type(str),
                )
            }
        )
    assert "Node 'a' contains a transition to non-existing node" in str(excinfo.value)


@pytest.mark.unit
def test_throws_when_self_reference() -> None:
    with pytest.raises(ValueError) as excinfo:
        LinePrefixParser(
            {
                "a": LinePrefixParserNode(
                    prefix="A:",
                    is_start=True,
                    is_end=True,
                    next=["a"],
                    field=ParserField.from_type(str),
                )
            }
        )
    assert "Node 'a' cannot point to itself." in str(excinfo.value)
